chained/water_version_control.go

```diff
@@ -0,0 +1,274 @@
+		return log.Errorf("failed to load history: %w", err)
+	}
+
+	if err := vc.updateLastTimeLoaded(transport); err != nil {
+		return log.Errorf("failed to update last time loaded: %w", err)
+	}
+
+	if err := vc.deleteOutdatedWASMFiles(); err != nil {
+		return log.Errorf("failed to delete outdated WASM files: %w", err)
+	}
+	return nil
+}
+
+func (vc *versionControl) updateLastTimeLoaded(transport string) error {
+	for i, h := range vc.history {
+		if h.Transport == transport {
+			vc.history[i].LastTimeLoaded = time.Now()
+			if err := vc.storeHistory(); err != nil {
+				return log.Errorf("failed to store history: %w", err)
+			}
+			return nil
+		}
+	}
+
+	vc.history = append(vc.history, history{
+		Transport:      transport,
+		LastTimeLoaded: time.Now(),
+	})
+	if err := vc.storeHistory(); err != nil {
+		return log.Errorf("failed to store history: %w", err)
+	}
+	return nil
+}
+
+func (vc *versionControl) loadHistory() error {
+	historyCSV := filepath.Join(vc.dir, "history.csv")
+
+	// if there's no history available, just initialize history
+	if _, err := os.Stat(historyCSV); os.IsNotExist(err) {
+		vc.history = []history{}
+		return nil
+	}
+
+	f, err := os.Open(historyCSV)
+	if err != nil {
+		return log.Errorf("failed to open file %s: %w", historyCSV, err)
+	}
+	defer f.Close()
+
+	history := []history{}
+	if err := gocsv.UnmarshalFile(f, &history); err != nil {
+		return log.Errorf("failed to unmarshal file %s: %w", historyCSV, err)
+	}
+
+	vc.history = history
+	return nil
+}
+
+func (vc *versionControl) deleteOutdatedWASMFiles() error {
+	affectedTransports := make([]string, 0)
+	for _, h := range vc.history {
+		// check if the file is outdated
+		sevenDaysAgo := time.Now().AddDate(0, 0, -7)
+		if h.LastTimeLoaded.Before(sevenDaysAgo) {
+			// delete the file
+			info := vc.wasmFilesAvailable[h.Transport]
+			if err := os.Remove(info.path); err != nil {
+				return log.Errorf("failed to delete file %s: %w", h.Transport, err)
+			}
+			delete(vc.wasmFilesAvailable, h.Transport)
+			affectedTransports = append(affectedTransports, h.Transport)
+		}
+	}
+
+	// remove the affected transports from the history
+	newHistory := make([]history, 0)
+	for _, h := range vc.history {
+		if !slices.Contains(affectedTransports, h.Transport) {
+			newHistory = append(newHistory, h)
+		}
+	}
+	vc.history = newHistory
+
+	if err := vc.storeHistory(); err != nil {
+		return log.Errorf("failed to store history: %w", err)
+	}
+	return nil
+}
+
+func (vc *versionControl) storeHistory() error {
+	historyCSV := filepath.Join(vc.dir, "history.csv")
+	f, err := os.Create(historyCSV)
+	if err != nil {
+		return log.Errorf("failed to create file %s: %w", historyCSV, err)
+	}
+	defer f.Close()
+
+	if err := gocsv.MarshalFile(&vc.history, f); err != nil {
+		return log.Errorf("failed to marshal file %s: %w", historyCSV, err)
+	}
+
+	return nil
+}
+
+func (vc *versionControl) downloadWASM(ctx context.Context, transport string, urls []string) (wasmInfo, error) {
```

Comment:
What if you gave it a `WASMDownloader`?  Then `versionControl` can just focus on version control stuff and doesn't need to worry about how the download actually works (e.g. the chained-then-direct-then-fronted stuff).

Reply:
Likewise, `GetWASM` (which calls this function) could accept a `WASMDownloader`.